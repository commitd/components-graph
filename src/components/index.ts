export * from './Graph'
export * from './GraphDebugControl'
export * from './GraphToolbar'
export * from './NodeViewer'
