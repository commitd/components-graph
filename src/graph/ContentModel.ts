import { v4 } from 'uuid'
import {
  ModelAttributeSet,
  ModelEdge,
  ModelGraphData,
  ModelNode,
} from './types'

export class ContentModel {
  public static fromRaw(data: ModelGraphData): ContentModel {
    let model = Object.values(data.nodes).reduce(
      (acc, node) => acc.addNode(node),
      ContentModel.createEmpty()
    )
    model = Object.values(data.edges).reduce(
      (acc, edge) => acc.addEdge(edge),
      model
    )
    return model
  }

  public static createEmpty(): ContentModel {
    return new ContentModel({}, {})
  }

  constructor(
    readonly nodes: Record<string, ModelNode>,
    readonly edges: Record<string, ModelEdge>
  ) {
    this.nodes = nodes
    this.edges = edges
  }

  containsNode(id: string): boolean {
    return this.getNode(id) != null
  }

  containsEdge(id: string): boolean {
    return this.getEdge(id) != null
  }

  getNode(id: string): ModelNode | undefined {
    return this.nodes[`${id}`]
  }

  private getExistingNode(id: string): ModelNode {
    const node = this.getNode(id)
    if (node == null) {
      throw new Error(`Node [${id}] does not exist`)
    }
    return node
  }

  getEdgesLinkedToNode(nodeId: string): ModelEdge[] {
    return Object.values(this.edges).filter(
      (e) => e.target === nodeId || e.source === nodeId
    )
  }

  addNode(node: Partial<ModelNode>): ContentModel {
    if (node.id != null && this.containsNode(node.id)) {
      throw new Error(`Cannot add node already in graph (${node.id})`)
    }
    const newNode: ModelNode = {
      id: node.id ?? v4(),
      attributes: node.attributes ?? {},
      ...node,
    }
    const nodes = { ...this.nodes, [newNode.id]: newNode }
    return new ContentModel(nodes, this.edges)
  }

  editNode(node: ModelNode): ContentModel {
    this.getExistingNode(node.id)
    return new ContentModel(
      { ...this.nodes, ...{ [node.id]: node } },
      this.edges
    )
  }

  addNodeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    const node = this.getExistingNode(id)
    const newAttributes = {
      ...node.attributes,
      ...{ [attributeName]: attributeValue },
    }
    const changedNode = {
      ...this.getExistingNode(id),
      ...{ attributes: newAttributes },
    }
    return this.editNode(changedNode)
  }

  /**
   * Remove node and associated edges
   * @param id id of the node
   */
  removeNode(id: string): ContentModel {
    const node = this.getNode(id)
    if (node == null) {
      return this
    }
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    let content: ContentModel = this
    this.getEdgesLinkedToNode(id).forEach((e) => {
      content = content.removeEdge(e.id)
    })
    return content.removeOrphanNode(id)
  }

  private removeOrphanNode(id: string): ContentModel {
    const withoutNode = { ...this.nodes }
    delete withoutNode[`${id}`]
    return new ContentModel(withoutNode, this.edges)
  }

  editNodeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    if (this.getExistingNode(id).attributes[`${attributeName}`] == null) {
      throw new Error(`Node [${id}] does not have attribute ${attributeName}`)
    }
    return this.addNodeAttribute(id, attributeName, attributeValue)
  }

  removeNodeAttribute(id: string, attributeName: string): ContentModel {
    const node = this.getExistingNode(id)
    const remainingAttributes = { ...node.attributes }
    delete remainingAttributes[`${attributeName}`]
    const n = { ...node, ...{ attributes: remainingAttributes } }
    return this.editNode(n)
  }

  addEdge(
    edge: Omit<ModelEdge, 'id' | 'attributes'> & {
      id?: string
      attributes?: ModelAttributeSet
    }
  ): ContentModel {
    // check source and target exist
    this.getExistingNode(edge.source)
    this.getExistingNode(edge.target)
    if (edge.id != null && this.containsEdge(edge.id)) {
      throw new Error(`Cannot add edge already in graph (${edge.id})`)
    }
    const newEdge: ModelEdge = {
      id: edge.id ?? v4(),
      attributes: edge.attributes ?? {},
      ...edge,
    }
    const edges = { ...this.edges, [newEdge.id]: newEdge }
    return new ContentModel(this.nodes, edges)
  }

  getEdge(id: string): ModelEdge | undefined {
    return this.edges[`${id}`]
  }

  private getExistingEdge(id: string): ModelEdge {
    const edge = this.getEdge(id)
    if (edge == null) {
      throw new Error(`Edge [${id}] does not exist`)
    }
    return edge
  }

  editEdge(edge: ModelEdge): ContentModel {
    // ensure edge exists
    this.getExistingEdge(edge.id)
    return new ContentModel(this.nodes, {
      ...this.edges,
      ...{ [edge.id]: edge },
    })
  }

  addEdgeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    const edge = this.getExistingEdge(id)
    const newAttributes = {
      ...edge.attributes,
      ...{ [attributeName]: attributeValue },
    }
    const changedEdge = {
      ...this.getExistingEdge(id),
      ...{ attributes: newAttributes },
    }
    return this.editEdge(changedEdge)
  }

  editEdgeAttribute<V>(
    id: string,
    attributeName: string,
    attributeValue: V
  ): ContentModel {
    if (this.getExistingEdge(id).attributes[`${attributeName}`] == null) {
      throw new Error(`Edge [${id}] does not have attribute ${attributeName}`)
    }
    return this.addEdgeAttribute(id, attributeName, attributeValue)
  }

  removeEdgeAttribute(id: string, attributeName: string): ContentModel {
    const edge = this.getExistingEdge(id)
    const remainingAttributes = { ...edge.attributes }
    delete remainingAttributes[`${attributeName}`]
    const e = { ...edge, ...{ attributes: remainingAttributes } }
    return this.editEdge(e)
  }

  removeEdge(id: string): ContentModel {
    const edge = this.getEdge(id)
    if (edge == null) {
      return this
    }
    const withoutEdge = { ...this.edges }
    delete withoutEdge[`${id}`]
    return new ContentModel(this.nodes, withoutEdge)
  }
}
