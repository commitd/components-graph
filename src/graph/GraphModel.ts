import { ContentModel } from './ContentModel'
import { DecoratorModel } from './DecoratorModel'
import { LayoutModel } from './LayoutModel'
import { SelectionModel } from './SelectionModel'
import {
  DecoratedEdge,
  DecoratedNode,
  GraphCommand,
  ModelEdge,
  ModelNode,
} from './types'

/** The GraphModel is a declarative, immutable definition of graph state
 
The GraphModel defines:
- Nodes/edges (ContentModel)
	- No requirement for ontology, typing etc
	- Attributes
- Decoration (DecoratorModel)
	- May define custom node/edge Decorators based on any property on the node/edge
- Selection (SelectionModel)
	- Multi-select
- Graph layout (LayoutModel)
	- Force Layout / Cola JS
	- Circle
	- Grid
	- Custom (supplying a function)
- CommandQueue
	- Trigger behaviour from outside the GraphModel/React component
	- Current commands: zoom in/out, relayout, refit
 */
export class GraphModel {
  private readonly contentModel: ContentModel
  private readonly selectionModel: SelectionModel
  private readonly decoratorModel: DecoratorModel
  private readonly layoutModel: LayoutModel
  private readonly commandQueue: GraphCommand[]

  constructor(
    contentModel: ContentModel,
    options: {
      selectionModel?: SelectionModel
      decoratorModel?: DecoratorModel
      layoutModel?: LayoutModel
      commandQueue?: GraphCommand[]
    } = {}
  ) {
    this.contentModel = contentModel
    this.decoratorModel =
      options.decoratorModel ?? DecoratorModel.createDefault()
    this.layoutModel = options.layoutModel ?? LayoutModel.createDefault()
    this.selectionModel =
      options.selectionModel ?? SelectionModel.createDefault()
    this.commandQueue = options.commandQueue ?? []
  }

  static applyContent(
    graphModel: GraphModel,
    contentModel: ContentModel
  ): GraphModel {
    return new GraphModel(contentModel, {
      ...graphModel.getOptions(),
      selectionModel: GraphModel.reconcileSelection(
        contentModel,
        graphModel.getSelection()
      ),
    })
  }

  static applySelection(
    graphModel: GraphModel,
    selectionModel: SelectionModel
  ): GraphModel {
    return new GraphModel(graphModel.getCurrentContent(), {
      ...graphModel.getOptions(),
      selectionModel: GraphModel.reconcileSelection(
        graphModel.getCurrentContent(),
        selectionModel
      ),
    })
  }

  static applyLayout(
    graphModel: GraphModel,
    layoutModel: LayoutModel
  ): GraphModel {
    return new GraphModel(graphModel.contentModel, {
      ...graphModel.getOptions(),
      layoutModel,
      selectionModel: graphModel.getSelection(),
    })
  }

  static createEmpty(): GraphModel {
    return new GraphModel(ContentModel.createEmpty())
  }

  static createWithContent(contentModel: ContentModel): GraphModel {
    return new GraphModel(contentModel)
  }

  getCurrentContent(): ContentModel {
    return this.contentModel
  }

  getDecorators(): DecoratorModel {
    return this.decoratorModel
  }

  // get the current fully decorated nodes
  get nodes(): DecoratedNode[] {
    return this.decoratorModel.getDecoratedNodes(
      Object.values(this.contentModel.nodes)
    )
  }

  // get the current fully decorated edges
  get edges(): DecoratedEdge[] {
    return this.decoratorModel.getDecoratedEdges(
      Object.values(this.contentModel.edges)
    )
  }

  get selectedNodes(): DecoratedNode[] {
    return this.decoratorModel.getDecoratedNodes(
      Array.from(this.getSelection().nodes)
        .map((n) => this.contentModel.getNode(n))
        .filter((n) => n != null)
        .map((n) => n as ModelNode)
    )
  }

  get selectedEdges(): DecoratedEdge[] {
    return this.decoratorModel.getDecoratedEdges(
      Array.from(this.getSelection().edges)
        .map((e) => this.contentModel.getEdge(e))
        .filter((e) => e != null)
        .map((e) => e as ModelEdge)
    )
  }

  getNode(id: string): DecoratedNode | undefined {
    const node = this.contentModel.getNode(id)
    if (node == null) {
      return node
    }
    return this.decoratorModel.getDecoratedNodes([node])[0]
  }

  getEdge(id: string): DecoratedEdge | undefined {
    const edge = this.contentModel.getEdge(id)
    if (edge == null) {
      return edge
    }
    return this.decoratorModel.getDecoratedEdges([edge])[0]
  }

  getCurrentLayout(): LayoutModel {
    return this.layoutModel
  }

  getSelection(): SelectionModel {
    return this.selectionModel
  }

  getCommands(): GraphCommand[] {
    return this.commandQueue
  }

  pushCommand(command: GraphCommand): GraphModel {
    return new GraphModel(this.contentModel, {
      ...this.getOptions(),
      commandQueue: [...this.commandQueue, command],
    })
  }

  clearCommands(): GraphModel {
    if (this.commandQueue.length === 0) {
      return this
    }
    return new GraphModel(this.contentModel, {
      ...this.getOptions(),
      commandQueue: [],
    })
  }

  private getOptions(): {
    selectionModel: SelectionModel
    decoratorModel: DecoratorModel
    layoutModel: LayoutModel
    commandQueue: GraphCommand[]
  } {
    return {
      selectionModel: this.selectionModel,
      decoratorModel: this.decoratorModel,
      layoutModel: this.layoutModel,
      commandQueue: this.commandQueue,
    }
  }

  private static reconcileSelection(
    contentModel: ContentModel,
    selectionModel: SelectionModel
  ): SelectionModel {
    const nodes = Array.from(selectionModel.nodes).filter((n) =>
      contentModel.containsNode(n)
    )
    const edges = Array.from(selectionModel.edges).filter((n) =>
      contentModel.containsEdge(n)
    )
    return new SelectionModel(nodes, edges)
  }
}
