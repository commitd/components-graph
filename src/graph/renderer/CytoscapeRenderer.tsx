/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
import { useTheme } from '@committed/components'
import {
  Css,
  EdgeCollection,
  EdgeDataDefinition,
  EdgeSingular,
  EventObject,
  LayoutOptions,
  NodeCollection,
  NodeDataDefinition,
  Stylesheet,
  use,
} from 'cytoscape'
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
//@ts-ignore
import ccola from 'cytoscape-cola'
import React, {
  CSSProperties,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'
import CytoscapeComponent from 'react-cytoscapejs'
import tinycolor from 'tinycolor2'
import { useDebouncedCallback } from 'use-debounce'
import { GraphModel } from '../GraphModel'
import { circle } from '../layouts/Circle'
import { cola } from '../layouts/Cola'
import { forceDirected } from '../layouts/ForceDirected'
import { grid } from '../layouts/Grid'
import { SelectionModel } from '../SelectionModel'
import {
  EdgeDecoration,
  GraphLayout,
  GraphRenderer,
  GraphRendererOptions,
  NodeDecoration,
  PresetGraphLayout,
} from '../types'
import {
  CustomLayoutOptions,
  CytoscapeGraphLayoutAdapter,
} from './CytoscapeGraphLayoutAdapter'
import { useCyListener } from './useCyListener'

/**
 * Call to initialize the additional modules.
 */
/*
 * This is a work-around - the modules do not seem to be registered when using as a library so must be called by the lib user.
 * Should be able to remove but may be a bundling issue.
 */
export function initializeCytoscape(
  cyuse: (module: cytoscape.Ext) => void
): void {
  try {
    cyuse(CytoscapeGraphLayoutAdapter.register)
    cyuse(ccola)
  } catch {
    // Ignore multiple attempts to initialize
  }
}

initializeCytoscape(use)

export interface CyGraphRendererOptions extends GraphRendererOptions {
  renderOptions: Pick<
    React.ComponentProps<typeof CytoscapeComponent>,
    | 'zoom'
    | 'minZoom'
    | 'maxZoom'
    | 'pan'
    | 'zoomingEnabled'
    | 'userZoomingEnabled'
    | 'boxSelectionEnabled'
    | 'autoungrabify'
    | 'autounselectify'
    | 'wheelSensitivity'
  >
}

const dblClick = () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let clicked: any | null = null
  return (evt: EventObject) => {
    if (clicked && clicked === evt.target) {
      clicked = null
      evt.preventDefault()
      evt.stopPropagation()
      evt.target.emit('dblclick', [evt])
    } else {
      clicked = evt.target
      setTimeout(() => {
        if (clicked && clicked === evt.target) {
          clicked = null
          evt.target.emit('dblclick:timeout', [evt])
        }
      }, 500)
    }
  }
}

const toNodeCyStyle = (d: Partial<NodeDecoration>): Css.Node | undefined => {
  const s: Css.Node = {
    backgroundColor: d.color,
    width: d.size,
    height: d.size,
    shape: d.shape,
    'background-image': d.icon,
    'background-opacity': d.opacity,
    'background-image-opacity': d.opacity,
    'border-opacity': d.opacity,
    'border-color': d.strokeColor,
  }
  if (d.label !== undefined) {
    s.label = d.label
  }
  if (Object.values(s).every((v) => v == null)) {
    return undefined
  } else {
    return s
  }
}

const toEdgeCyStyle = (e: Partial<EdgeDecoration>): Css.Edge | undefined => {
  const s: Css.Edge = {
    'line-color': e.color,
    'target-arrow-color': e.color,
    opacity: e.opacity,
  }

  if (e.label !== undefined) {
    s.label = e.label
  }
  if (Object.values(s).every((v) => v == null)) {
    return undefined
  } else {
    return s
  }
}

const Renderer: GraphRenderer<CyGraphRendererOptions>['render'] = ({
  graphModel,
  onChange,
  onViewNode,
  options,
}) => {
  const layouts: Record<PresetGraphLayout | 'custom', LayoutOptions> = {
    'force-directed': forceDirected,
    circle,
    grid,
    cola,
    custom: {
      name: CytoscapeGraphLayoutAdapter.LAYOUT_NAME,
      model: graphModel,
      algorithm: graphModel.getCurrentLayout().getLayout(),
    } as CustomLayoutOptions & LayoutOptions,
  }
  const nodes = graphModel.nodes
  const edges = graphModel.edges
  const selection = graphModel.getSelection()
  const [cytoscape, setCytoscape] = useState<cytoscape.Core>()
  const layoutStart = useRef<number>()
  const [theme, resolveThemeToken] = useTheme()
  const dirty = graphModel.getCurrentLayout().isDirty()
  const layout = graphModel.getCurrentLayout().getLayout()

  // prevent rapid selection updates from fighting each other
  // and improve performance overheads by debouncing selection updates

  const triggerLayout: (layout: GraphLayout) => void = useDebouncedCallback(
    (graphLayout: GraphLayout) => {
      if (cytoscape != null) {
        const l =
          typeof graphLayout === 'string'
            ? layouts[graphLayout]
            : ({
                name: CytoscapeGraphLayoutAdapter.LAYOUT_NAME,
                model: graphModel,
                algorithm: graphLayout,
              } as CustomLayoutOptions)
        if (l == null) {
          throw new Error('Layout does not exist')
        }
        cytoscape.layout(l as LayoutOptions).run()
      }
    },
    200
  )

  const pendingSelection = useRef<SelectionModel>(selection)

  useEffect(() => {
    pendingSelection.current = selection
  }, [selection])
  const selectionUpdate = useDebouncedCallback(() => {
    if (pendingSelection.current != null) {
      onChange(GraphModel.applySelection(graphModel, pendingSelection.current))
    }
  }, 100)

  const updateSelection = useCallback(
    (updater: (s: SelectionModel) => SelectionModel) => {
      pendingSelection.current = updater(pendingSelection.current)
      selectionUpdate()
    },
    [pendingSelection, selectionUpdate]
  )
  const handleViewNode = useCallback(
    (id: string) => {
      const node = graphModel.getNode(id)
      if (node != null && onViewNode != null) {
        onViewNode(node)
      }
    },
    [onViewNode, graphModel]
  )

  const updateLayout = useCallback(() => {
    triggerLayout(layout)
  }, [triggerLayout, layout])
  const selectNode = useCallback(
    (e: EventObject) => {
      const selectedNodes = e.target as NodeCollection
      updateSelection((s) => s.addNodes(selectedNodes.map((n) => n.id())))
    },
    [updateSelection]
  )
  const unselectNode = useCallback(
    (e: EventObject) => {
      const selectedNodes = e.target as NodeCollection
      updateSelection((s) => s.removeNodes(selectedNodes.map((n) => n.id())))
    },
    [updateSelection]
  )
  const selectEdge = useCallback(
    (e: EventObject) => {
      const selectedEdges = e.target as EdgeCollection
      updateSelection((s) => s.addEdges(selectedEdges.map((n) => n.id())))
    },
    [updateSelection]
  )
  const unselectEdge = useCallback(
    (e: EventObject) => {
      const selectedEdges = e.target as EdgeCollection
      updateSelection((s) => s.removeEdges(selectedEdges.map((n) => n.id())))
    },
    [updateSelection]
  )
  const layoutStarting = useCallback(() => {
    console.debug('Layout started')
    layoutStart.current = Date.now()
  }, [])
  const layoutStopping = useCallback(() => {
    if (layoutStart.current != null) {
      console.debug(`Layout took ${Date.now() - layoutStart.current}ms`)
    }
  }, [])

  useCyListener(cytoscape, updateLayout, 'add remove', 'edge')
  useCyListener(cytoscape, updateLayout, 'resize', 'edge')
  useCyListener(cytoscape, selectNode, 'select', 'node')
  useCyListener(cytoscape, unselectNode, 'unselect', 'node')
  useCyListener(cytoscape, selectEdge, 'select', 'edge')
  useCyListener(cytoscape, unselectEdge, 'unselect', 'edge')
  useCyListener(cytoscape, layoutStarting, 'layoutstart')
  useCyListener(cytoscape, layoutStopping, 'layoutstop')
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const dblClickCallback = useCallback(dblClick(), [])

  const dblclickNode = useCallback(
    (e: EventObject) => {
      const selectedNodes = e.target as NodeCollection
      const node = selectedNodes[0]
      if (node != null) {
        handleViewNode(node.id())
      }
    },
    [handleViewNode]
  )
  useCyListener(cytoscape, dblClickCallback, 'click', 'node')
  useCyListener(cytoscape, dblclickNode, 'dblclick', 'node')

  useEffect(() => {
    if (dirty) {
      triggerLayout(layout)
      onChange(
        GraphModel.applyLayout(
          graphModel,
          graphModel.getCurrentLayout().validate()
        )
      )
    }
  }, [dirty, layout, onChange, graphModel, triggerLayout])

  const commands = graphModel.getCommands()
  useEffect(() => {
    if (cytoscape == null) {
      return
    }
    const zoomAmount = 1 + (options.renderOptions.wheelSensitivity ?? 1)
    const position = { x: cytoscape.width() / 2, y: cytoscape.height() / 2 }
    commands.forEach((c) => {
      switch (c.type) {
        case 'ZoomIn':
          cytoscape.zoom({
            level: cytoscape.zoom() * zoomAmount,
            position,
          })
          break
        case 'ZoomOut':
          cytoscape.zoom({
            level: cytoscape.zoom() / zoomAmount,
            position,
          })
          break
        case 'Layout':
          triggerLayout(layout)
          break
        case 'Refit':
          cytoscape.fit()
          break
        default:
          console.warn('Unsupported command')
      }
    })
    onChange(graphModel.clearCommands())
    // TODO check for correct dependency list
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [commands])

  const resolveThemeValue = useCallback(
    <T extends string | number>(val: T): T => {
      if (val === undefined) {
        return val
      }

      if (typeof val === 'string') {
        const resolved = resolveThemeToken(val)
        if (resolved.startsWith('hsl')) {
          return `#${tinycolor(resolved).toHex()}` as T
        } else {
          return resolved as T
        }
      }
      return val
    },
    [resolveThemeToken]
  )

  const resolveThemedObject = useCallback(
    // eslint-disable-next-line @typescript-eslint/ban-types
    <T extends object>(obj: T): T => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return Object.keys(obj).reduce(function (result, key) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call
        if (obj[key] !== undefined) result[key] = resolveThemeValue(obj[key])
        return result
      }, {} as T)
    },
    [resolveThemeValue]
  )

  const elements = useMemo(
    () => [
      ...nodes.map((n) => {
        const node: NodeDataDefinition = {
          id: n.id,
          label: n.label,
          model: n,
        }
        const { label, ...style } = resolveThemedObject(
          toNodeCyStyle(n.getDecorationOverrides()) ?? {}
        )
        return {
          data: { ...node, label },
          style,
          selected: selection.nodes.has(n.id),
        }
      }),
      ...edges.map((e) => {
        const edge: EdgeDataDefinition = {
          id: e.id,
          source: e.source,
          target: e.target,
          label: e.label,
          selected: selection.edges.has(e.id),
          model: e,
        }
        const { label, ...style } = resolveThemedObject(
          toEdgeCyStyle(e.getDecorationOverrides()) ?? {}
        )
        return {
          data: { ...edge, label },
          style,
          classes: 'bezier',
        }
      }),
    ],
    [nodes, edges, resolveThemedObject, selection.nodes, selection.edges]
  )

  const defaultNodeStyles: Stylesheet = useMemo(() => {
    const nodeDefaults = graphModel.getDecorators().getDefaultNodeDecorator()()
    return {
      selector: 'node',
      style: resolveThemedObject({
        label: 'data(label)',
        // label styles
        'text-background-color': '$colors$paper',
        color: '$colors$text',
        'text-background-opacity': 0.7,
        'text-margin-y': -4,
        'text-background-shape': 'roundrectangle',

        // icon styles
        'background-width': '100%',
        'background-height': '100%',
        'border-width': (e) => {
          const strokeSize = graphModel.getNode(e.id())?.strokeSize
          return strokeSize == null
            ? nodeDefaults.strokeSize
            : e.selected()
            ? strokeSize * 3
            : strokeSize
        },
        ...toNodeCyStyle(nodeDefaults),
      }),
    }
  }, [graphModel, resolveThemedObject])

  const defaultEdgeStyles: Stylesheet = useMemo(() => {
    const edgeDefaults = graphModel.getDecorators().getDefaultEdgeDecorator()()
    return {
      selector: 'edge',
      style: resolveThemedObject({
        label: 'data(label)',
        // label styles
        'text-background-color': '$colors$paper',
        color: '$colors$text',
        'text-background-opacity': 0.7,
        'text-margin-y': -4,
        'text-background-shape': 'roundrectangle',
        'target-arrow-shape': 'triangle',
        'target-endpoint': 'outside-to-node-or-label',
        width: (e: EdgeSingular) => {
          const size = graphModel.getEdge(e.id())?.size
          return size == null
            ? edgeDefaults.size
            : e.selected()
            ? size * 3
            : size
        },
        ...toEdgeCyStyle(edgeDefaults),
      }),
    }
  }, [graphModel, resolveThemedObject])

  if (theme === undefined) {
    // Wait for theme to be defined
    return null
  }

  return (
    <CytoscapeComponent
      elements={elements}
      style={{
        width:
          options.width == null || options.width === 'full-width'
            ? '100%'
            : options.width,
        height: options.height === 'full-height' ? '100%' : options.height,
        backgroundColor: resolveThemeValue(
          '$colors$paper'
        ) as CSSProperties['backgroundColor'],
        overflow: 'hidden',
      }}
      cy={(cyCore): void => {
        setCytoscape(cyCore)
      }}
      stylesheet={[defaultNodeStyles, defaultEdgeStyles]}
      {...options.renderOptions}
    />
  )
}

export const cytoscapeRenderer: GraphRenderer<CyGraphRendererOptions> = {
  render: Renderer,
  defaultOptions: {
    height: 300,
    renderOptions: {
      maxZoom: 5,
      minZoom: 0.05,
    },
  },
  layouts: ['circle', 'cola', 'force-directed', 'grid'],
}
