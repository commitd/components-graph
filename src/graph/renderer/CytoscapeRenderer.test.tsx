import { GraphModel } from '../GraphModel'
import { cytoscapeRenderer } from './CytoscapeRenderer'
import { renderLight } from '../../setupTests'
import React from 'react'
import { Graph } from '../../components'
import { ModelNode, ModelEdge } from '..'
import { ContentModel } from '../ContentModel'

let cytoscape: any

jest.mock('react-cytoscapejs', () => {
  return {
    __esModule: true,
    default: ({ elements, cy }: { elements: any; cy: (cy: any) => void }) => {
      React.useEffect(() => {
        cy(cytoscape)
      }, [])

      return <div>{JSON.stringify(elements)}</div>
    },
  }
})

// Remove the debounce, to simplify tests
jest.mock('use-debounce', () => {
  return {
    useDebouncedCallback: (fn: any) => ({
      callback: fn,
    }),
  }
})

let graphModel: GraphModel

const node1: ModelNode = {
  id: 'node1',
  attributes: {
    employer: 'Committed',
  },
  color: 'yellow',
  label: 'Node 1',
  size: 10,
  strokeColor: 'black',
  opacity: 1,
  shape: 'ellipse',
  strokeSize: 2,
}

const node2: ModelNode = {
  id: 'node2',
  attributes: {
    employer: 'Government',
  },
  color: 'green',
  label: 'Node 2',
  size: 12,
  strokeColor: 'black',
  opacity: 0.9,
  shape: 'rectangle',
  strokeSize: 3,
}

const edge1: ModelEdge = {
  id: 'edge1',
  attributes: {
    role: 'client',
  },
  source: node1.id,
  target: node2.id,
}

beforeEach(() => {
  graphModel = GraphModel.applyContent(
    GraphModel.createEmpty(),
    ContentModel.createEmpty().addNode(node1).addNode(node2).addEdge(edge1)
  )
  cytoscape = {
    on: jest.fn(),
    fit: jest.fn(),
    zoom: jest.fn().mockImplementation(() => 1),
    removeAllListeners: jest.fn(),
    layout: jest.fn().mockImplementation(() => ({
      run: jest.fn(),
    })),
    width: () => 200,
    height: () => 400,
  }
})

it('can be rendered', () => {
  const onChange = jest.fn()

  const { asFragment, unmount } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )
  expect(asFragment()).toMatchSnapshot()
  expect(cytoscape.on).toHaveBeenCalledTimes(9)

  unmount()

  expect(cytoscape.removeAllListeners).toHaveBeenCalledTimes(1)
})

it('layout command triggers layout', () => {
  const onChange = jest.fn()

  const { rerender } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )

  graphModel = graphModel.pushCommand({ type: 'Layout' })

  cytoscape.layout.mockClear()

  rerender(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )

  expect(cytoscape.layout).toHaveBeenCalledTimes(2)
  expect(onChange).toBeCalledWith(graphModel.clearCommands())
})

it('refit command triggers refit', () => {
  const onChange = jest.fn()

  const { rerender } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )

  graphModel = graphModel.pushCommand({ type: 'Refit' })

  cytoscape.layout.mockClear()

  rerender(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )
  expect(cytoscape.fit).toHaveBeenCalledTimes(1)
  expect(onChange).toBeCalledWith(graphModel.clearCommands())
})

it('invalidate layout triggers layout', () => {
  const onChange = jest.fn()

  const { rerender } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )

  graphModel = GraphModel.applyLayout(
    graphModel,
    graphModel.getCurrentLayout().invalidate()
  )

  cytoscape.layout.mockClear()

  rerender(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )

  expect(onChange).toBeCalledWith(
    GraphModel.applyLayout(graphModel, graphModel.getCurrentLayout().validate())
  )
  expect(cytoscape.layout).toHaveBeenCalledTimes(1)
})

it('ZoomIn command triggers zoom', () => {
  const onChange = jest.fn()

  const { rerender } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )

  graphModel = graphModel.pushCommand({ type: 'ZoomIn' })

  rerender(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )
  expect(cytoscape.zoom).toHaveBeenCalledTimes(2)
  expect(onChange).toBeCalledWith(graphModel.clearCommands())
})

it('ZoomOut command triggers zoom', () => {
  const onChange = jest.fn()

  const { rerender } = renderLight(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )

  graphModel = graphModel.pushCommand({ type: 'ZoomOut' })

  rerender(
    <Graph
      model={graphModel}
      onModelChange={onChange}
      renderer={cytoscapeRenderer}
      options={{ height: 'full-height' }}
    />
  )
  expect(cytoscape.zoom).toHaveBeenCalledTimes(2)
  expect(onChange).toBeCalledWith(graphModel.clearCommands())
})
