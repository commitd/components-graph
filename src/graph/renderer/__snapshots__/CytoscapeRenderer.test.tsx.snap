// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered 1`] = `
<DocumentFragment>
  <div>
    [{"data":{"id":"node1","label":"Node 1","model":{"id":"node1","attributes":{"employer":"Committed"},"color":"yellow","label":"Node 1","size":10,"strokeColor":"black","opacity":1,"shape":"ellipse","strokeSize":2}},"style":{"backgroundColor":"yellow","width":10,"height":10,"shape":"ellipse","background-opacity":1,"background-image-opacity":1,"border-opacity":1,"border-color":"black"},"selected":false},{"data":{"id":"node2","label":"Node 2","model":{"id":"node2","attributes":{"employer":"Government"},"color":"green","label":"Node 2","size":12,"strokeColor":"black","opacity":0.9,"shape":"rectangle","strokeSize":3}},"style":{"backgroundColor":"green","width":12,"height":12,"shape":"rectangle","background-opacity":0.9,"background-image-opacity":0.9,"border-opacity":0.9,"border-color":"black"},"selected":false},{"data":{"id":"edge1","source":"node1","target":"node2","selected":false,"model":{"id":"edge1","attributes":{"role":"client"},"source":"node1","target":"node2"}},"style":{},"classes":"bezier"}]
  </div>
</DocumentFragment>
`;
